### PR TITLE
Resolve: [Deposit] Add business logic to check pending assets 

### LINKED ISSUE #1495: [Deposit] Add business logic to check pending assets

### PR DESCRIPTION
**Pull request by thatStrangeGuyThorchain**
- introduced manual loading data from `/thorchain/pool/{pool}/liquidity_providers`
- introduced custom `io-ts` codecs to validate responses
- passed data to the `DepositView`

https://user-images.githubusercontent.com/67793194/120650326-57a49e00-c486-11eb-8b6e-56b50f552040.mov

closes #1495

### CODE REVIEW

File: src/renderer/services/thorchain/types.ts [hunk 1]
```diff
@@ -70,3 +75,81 @@ export type Mimir = t.TypeOf<typeof MimirIO>
 
 export type MimirLD = LiveData<Error, Mimir>
 export type MimirRD = RD.RemoteData<Error, Mimir>
+
+export type GetLiquidityProvidersParams = {
+  asset: Asset
+  network: Network
+  assetDecimal: number
+}
```
Comment (veado) [inline src/renderer/services/thorchain/types.ts]: `{ assetWD: AssetWithDecimal, network: Network }`

File: src/renderer/services/thorchain/thorNode.ts [hunk 2]
```diff
@@ -103,6 +117,68 @@ const getNodeInfo$ = (node: Address, network: Network): NodeInfoLD =>
...
+        RxOp.map(A.filterMap(FP.flow(LiquidityProviderIO.decode, O.fromEither))),
+        RxOp.map(RD.success),
+        liveData.map(
+          A.map((provider) => ({
+            asset: provider.asset,
+            address: provider.rune_address,
+            pendingRune: baseAmount(provider.pending_rune, THORCHAIN_DECIMAL),
+            pendingAsset: baseAmount(provider.pending_asset, assetDecimal),
+            runeDepositValue: baseAmount(provider.rune_deposit_value, THORCHAIN_DECIMAL),
+            assetDepositValue: baseAmount(provider.asset_deposit_value, assetDecimal)
```
Comment (veado) [inline src/renderer/services/thorchain/thorNode.ts]: Do we need `runeDepositValue` + `assetDepositValue`?

File: src/renderer/services/thorchain/thorNode.ts [hunk 3]
```diff
@@ -103,6 +117,68 @@ const getNodeInfo$ = (node: Address, network: Network): NodeInfoLD =>
...
+  assetDecimal,
+  walletAddress
+}: GetLiquidityProviderParams): LiquidityProviderLD =>
+  FP.pipe(
+    getLiquidityProviders({
+      asset,
+      network,
+      assetDecimal
+    }),
+    liveData.map(A.findFirst((provider) => provider.address === walletAddress))
```
Comment (veado) [inline src/renderer/services/thorchain/thorNode.ts]: AFAIK any LP can be twice in that list - as a sym. and asym. LP.

File: src/renderer/services/thorchain/types.ts [hunk 4]
```diff
@@ -70,3 +75,81 @@ export type Mimir = t.TypeOf<typeof MimirIO>
...
+      stringAsset,
+      assetFromString,
+      E.fromNullable(new Error('Invalid asset')),
+      E.fold((e) => D.failure(stringAsset, e.message), D.success)
+    )
+  )
+)
+
+/**
+ * Custom type validator to check if string is Asset instanse
```
Comment (veado) [inline src/renderer/services/thorchain/types.ts]: [typo] `instance`

File: src/renderer/services/thorchain/types.ts [hunk 5]
```diff
@@ -70,3 +75,81 @@ export type Mimir = t.TypeOf<typeof MimirIO>
...
+  (input): input is Asset => {
+    if (typeof input === 'string') {
+      const asset = assetFromString(input)
+
+      if (asset) {
+        return isValidAsset(asset)
+      }
+    }
+    return false
+  },
```
Comment (veado) [inline src/renderer/services/thorchain/types.ts:130]: Why do we need two checks for `input === string` (here ^ and in next step)?

File: src/renderer/views/deposit/DepositView.tsx [hunk 6]
```diff
@@ -143,6 +149,42 @@ export const DepositView: React.FC<Props> = () => {
...
+              ([{ asset, decimal }, address]) =>
+                getLiquidityProvider({ asset, network, walletAddress: address, assetDecimal: decimal })
+            )
+          )
+        )
+      ),
+    RD.initial
+  )
+
+  const _hasMissingDeposits: boolean = useMemo(
```
Comment (veado) [inline src/renderer/views/deposit/DepositView.tsx]: ^ Still needed ?

**Review by veado — CHANGES_REQUESTED** [review]

File: src/renderer/services/thorchain/types.ts [hunk 7]
```diff
@@ -70,3 +75,81 @@ export type Mimir = t.TypeOf<typeof MimirIO>
...
+      )
+    }
+    return t.failure(input, context)
+  },
+  assetDecoder.decode
+)
+
+export const LiquidityProviderIO = t.type({
+  // last_add_height: t.number,
+  // units: t.string,
```
Comment (veado) [inline src/renderer/services/thorchain/types.ts]: ^ still needed?

File: src/renderer/services/thorchain/thorNode.ts (hunk 2 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/services/thorchain/thorNode.ts]: no. left this for possible future needs. can remove that. np

File: src/renderer/services/thorchain/types.ts (hunk 5 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/services/thorchain/types.ts:130]: cuz initially `input` has `unknown` type and we strictly check for this value to be a string. the main idea of all codecs of `io-ts` is that you always getting `unknown` value and do all the checks you need by hands

File: src/renderer/services/thorchain/types.ts (hunk 7 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/services/thorchain/types.ts]: left this for future possible needs just to have an easy and fast way to understand what's also inside of thee response. can remove if you want

File: src/renderer/views/deposit/DepositView.tsx (hunk 6 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/views/deposit/DepositView.tsx]: for the next PR about UI for missing deposits

File: src/renderer/services/thorchain/thorNode.ts (hunk 2 above)
Comment (veado) [reply src/renderer/services/thorchain/thorNode.ts]: Yep, let's remove to have things we need only.

File: src/renderer/services/thorchain/types.ts (hunk 5 above)
Comment (veado) [reply src/renderer/services/thorchain/types.ts:130]: Make sense.

File: src/renderer/views/deposit/DepositView.tsx (hunk 6 above)
Comment (veado) [reply src/renderer/views/deposit/DepositView.tsx]: Ok, pls add a comment about it.

File: src/renderer/services/thorchain/types.ts (hunk 7 above)
Comment (veado) [reply src/renderer/services/thorchain/types.ts]: Let's remove it.

**Review by veado — DISMISSED** [review]
Nice one :+1:  _Just a small clean-up needed and ready to merge it then._

**Pull request merged**